src/Happstack/Server/FileServe/BuildingBlocks.hs

```diff
@@ -383,15 +385,43 @@ fileServe' serveFn mimeFn indexFn localPath = do
 -- Nothing
 -- >>> combineSafe "/var/uploads/" "../uploads/home/../etc/passwd"
 -- Just "/var/uploads/etc/passwd"
-combineSafe :: FilePath -> FilePath -> Maybe FilePath
-combineSafe root path =
-    if commonPrefix [root', joined] == root'
-      then Just $ encodeString joined
-      else Nothing
+combineSafe :: MonadIO m => FilePath -> FilePath -> m (Maybe FilePath)
+combineSafe root path = do
+    root' <- liftIO $ makeAbsolute root
+    let path' = normalise path
+    pure $ do
+        combined <- combineReduce root' path'
+        if root' `isPrefixOf` combined
+            then Just combined
+            else Nothing
```

Comment:
instead of the nested do block i'd prefer:

```haskell
pure $
    case combineReduce root' path' of
         Just combined | root' `isPrefixOf` combined ->
             Just combined
         _ ->
             Nothing
```

Reply:
Ok, fixed

Reply:
I made a small error, but it's been corrected and the tests are passing again.